### PR TITLE
create wrapper deprecation

### LINKED ISSUE #9775: [Deprecation] InputConnectionCompat.createWrapper API deprecated
###### Reproduction Steps

1. update dependencies (androidx core-ktx #9725)
2. run a build
3. see build warning failing build

###### Expected Result

clean compile without ignoring deprecation

###### Actual Result

```
/home/mike/work/AnkiDroid/Anki-Android/AnkiDroid/src/main/java/com/ichi2/anki/FieldEditText.java:146: warning: [deprecation] createWrapper(InputConnection,EditorInfo,OnCommitContentListener) in InputConnectionCompat has been deprecated
        return InputConnectionCompat.createWrapper(inputConnection, editorInfo, (contentInfo, flags, opts) -> {
                                    ^
```

We stumbled on this deprecation:
https://developer.android.com/reference/androidx/core/view/inputmethod/InputConnectionCompat#createWrapper(android.view.inputmethod.InputConnection,%20android.view.inputmethod.EditorInfo,%20androidx.core.view.inputmethod.InputConnectionCompat.OnCommitContentListener)

###### Debug info
Refer to the [support page](https://ankidroid.org/docs/help.html) if you are unsure where to get the "debug info".

###### Research
*Enter an [x] character to confirm the points below:*

- [ ] I have read the [support page](https://ankidroid.org/docs/help.html) and am reporting a bug or enhancement request specific to AnkiDroid
- [ ] I have checked the [manual](https://ankidroid.org/docs/manual.html) and the [FAQ](https://github.com/ankidroid/Anki-Android/wiki/FAQ) and could not find a solution to my issue
- [ ] I have searched for similar existing issues here and on the user forum
- [ ] (Optional) I have confirmed the issue is not resolved in the latest alpha release ([instructions](https://docs.ankidroid.org/manual.html#betaTesting))

### PR DESCRIPTION
**Pull request by murugu-21**
## Purpose / Description
 Input connection createWrapper current implementation is deprecated in function onCreateInputConnection in AnkiDroid/Anki-Android/AnkiDroid/src/main/java/com/ichi2/anki/FieldEditText.java.

## Fixes
Fixes #9775 

## Approach

As suggested in [documentation](https://developer.android.com/reference/androidx/core/view/inputmethod/InputConnectionCompat#createWrapper(android.view.inputmethod.InputConnection,%20android.view.inputmethod.EditorInfo,%20androidx.core.view.inputmethod.InputConnectionCompat.OnCommitContentListener)), i have used createWrapper(View,InputConnection,EditorInfo) and ViewCompat.setOnReceiveContentListener with implementation of OnReceiveContentListener interface.

I have not reimplemented the request read uri permission part alone because its documentation states read uri permission is granted by default [here](https://developer.android.com/reference/androidx/core/view/OnReceiveContentListener#uri-permissions).

reference [implementation](https://developer.android.com/reference/androidx/core/view/OnReceiveContentListener) of onReceiveContentListener

## How Has This Been Tested?

local run of all androidtest in com.ichi2 output:
image

## Checklist
_Please, go through these checks before submitting the PR._

- [x] You have not changed whitespace unnecessarily (it makes diffs hard to read)
- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] Your code follows the style of the project (e.g. never omit braces in `if` statements) 
- [ ] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code
- [ ] UI changes: include screenshots of all affected screens (in particular showing any new or changed strings)
- [ ] UI Changes: You have tested your change using the [Google Accessibility Scanner](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor)

### CODE REVIEW

File: AnkiDroid/src/main/java/com/ichi2/anki/FieldEditText.java [hunk 1]
```diff
@@ -137,51 +140,48 @@ public void setImagePasteListener(ImagePasteListener imageListener) {
...
-                    return false;
-                }
-            }
+            Pair<ContentInfoCompat, ContentInfoCompat> split = payload.partition(
+                    item -> item.getUri() != null);
+            ContentInfoCompat uriContent = split.first;
+            ContentInfoCompat remaining = split.second;
 
-            ClipDescription description = contentInfo.getDescription();
+            ClipDescription description = remaining.getClip().getDescription();
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/anki/FieldEditText.java:161]: can `remaining`/`getClip()` be null? Do we need to guard against this?

File: AnkiDroid/src/main/java/com/ichi2/anki/FieldEditText.java [hunk 2]
```diff
@@ -137,51 +140,48 @@ public void setImagePasteListener(ImagePasteListener imageListener) {
...
-    @SuppressWarnings("deprecation") // Tracked in #9775
     public InputConnection onCreateInputConnection(EditorInfo editorInfo) {
         InputConnection inputConnection = super.onCreateInputConnection(editorInfo);
         if (inputConnection == null) {
             return null;
         }
         EditorInfoCompat.setContentMimeTypes(editorInfo, IMAGE_MIME_TYPES);
-        return InputConnectionCompat.createWrapper(inputConnection, editorInfo, (contentInfo, flags, opts) -> {
+
+        ViewCompat.setOnReceiveContentListener(this, IMAGE_MIME_TYPES, (view, payload)-> {
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/anki/FieldEditText.java:150]:
```suggestion
        ViewCompat.setOnReceiveContentListener(this, IMAGE_MIME_TYPES, (view, payload) -> {
```

File: AnkiDroid/src/main/java/com/ichi2/anki/FieldEditText.java [hunk 3]
```diff
@@ -137,51 +140,48 @@ public void setImagePasteListener(ImagePasteListener imageListener) {
...
             if (!ClipboardUtil.hasImage(description)) {
-                return false;
+                return null;
             }
 
-            try {
-                if (!onImagePaste(contentInfo.getContentUri())) {
-                    return false;
+            if (uriContent != null) {
+                Uri uri = uriContent.getClip().getItemAt(0).getUri();
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/anki/FieldEditText.java:168]: Is this guaranteed to be non-null?

File: AnkiDroid/src/main/java/com/ichi2/anki/FieldEditText.java [hunk 4]
```diff
@@ -137,51 +140,48 @@ public void setImagePasteListener(ImagePasteListener imageListener) {
...
             return null;
         }
         EditorInfoCompat.setContentMimeTypes(editorInfo, IMAGE_MIME_TYPES);
-        return InputConnectionCompat.createWrapper(inputConnection, editorInfo, (contentInfo, flags, opts) -> {
+
+        ViewCompat.setOnReceiveContentListener(this, IMAGE_MIME_TYPES, (view, payload)-> {
 
             if (mImageListener == null) {
-                return false;
+                return null;
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/anki/FieldEditText.java:153]: Has this been reversed?

From the old docs 

> `true` if this request is accepted by the application, no matter if the request is already handled or still being handled in background. `false` to use the default implementation

**Review by david-allison-1 — COMMENTED** [review]
Thanks so much! Have a few questions

File: AnkiDroid/src/main/java/com/ichi2/anki/FieldEditText.java (hunk 1 above)
Comment (murugu-21) [reply AnkiDroid/src/main/java/com/ichi2/anki/FieldEditText.java:161]: getClip is non nullable
image

**Comment by murugu-21** [comment]
can images be inserted in clipboard without content uri?. If not, my current implementation has to be changed.

File: AnkiDroid/src/main/java/com/ichi2/anki/FieldEditText.java (hunk 1 above)
Comment (murugu-21) [reply AnkiDroid/src/main/java/com/ichi2/anki/FieldEditText.java]: turns out remaining can be null and never contains images. I have rewritten the function based on this.

File: AnkiDroid/src/main/java/com/ichi2/anki/FieldEditText.java (hunk 4 above)
Comment (murugu-21) [reply AnkiDroid/src/main/java/com/ichi2/anki/FieldEditText.java]: I should have returned remaining here. new docs suggests returning remaining as it contains all other types of input without uri (in this case, text I suppose).

**Comment by david-allison-1** [comment]
> can images be inserted in clipboard without content uri?. If not, my current implementation has to be changed.

I don't know this one. We may be accepting file URIs from some keyboards.

The docs imply that `content://` will be the majority of complex clip events.

https://developer.android.com/reference/android/content/ClipData

**Comment by murugu-21** [comment]
> > can images be inserted in clipboard without content uri?. If not, my current implementation has to be changed.
> 
> I don't know this one. We may be accepting file URIs from some keyboards.
> 
> The docs imply that `content://` will be the majority of complex clip events.
> 
> https://developer.android.com/reference/android/content/ClipData
> 
> Edit:
> 
> > This usually contains a content provider URI, although any URI is allowed.
> 
> https://developer.android.com/guide/topics/text/copy-paste

[getUri()](https://developer.android.com/reference/android/content/ClipData.Item#getUri()) seems to return file uris too. ClipData.Item doesn't seem to have another method for retrieving images.

File: AnkiDroid/src/main/java/com/ichi2/anki/FieldEditText.java [hunk 5]
```diff
@@ -137,51 +141,47 @@ public void setImagePasteListener(ImagePasteListener imageListener) {
...
 
             if (!ClipboardUtil.hasImage(description)) {
-                return false;
+                return remaining;
             }
 
-            try {
-                if (!onImagePaste(contentInfo.getContentUri())) {
-                    return false;
+            for (int i = 0; i < clip.getItemCount(); i++){
```
Comment (david-allison-1) [inline AnkiDroid/src/main/java/com/ichi2/anki/FieldEditText.java:169]: optional: `@KotlinCleanup("add extension method to iterate clip items")`:

**Review by david-allison-1 — APPROVED** [review]
Looks good to me, we'll find out in beta/production if we're mistaken about the URIs

**Comment by mikehardy** [comment]
In your testing section, you say:

> local run of all androidtest in com.ichi2 output:

Okay, the automated tests are good, but have you actually *used* the feature to make sure it still works? Just curious, as the place to mention that would be in the test section and based on what you write, it appears you have not actually used the feature at all to make sure it still works

**Review by mikehardy — APPROVED** [review]
Seems fine pending confirmation that feature has actually been tested on device after the change vs just automated test

passes this locally

> ./gradlew --stop && gh pr checkout 10265 && uom && ./gradlew clean jacocoTestReport :api:lintRelease :AnkiDroid:lintPlayRelease ktlintCheck

**Comment by murugu-21** [comment]
From local testing, text paste seems to work properly. But pasting image from clipboard results in something went wrong toast. 

Through debugging I was able to gather that this exception occurs here:

[src/main/java/com/ichi2/anki/MediaRegistration.kt](https://github.com/ankidroid/Anki-Android/blob/dddac58078ee18236b9b703355bfd159f3311075/AnkiDroid/src/main/java/com/ichi2/anki/MediaRegistration.kt)

[AnkiDroidApp.sendExceptionReport("File is invalid issue:8880", "RegisterMediaForWebView:onImagePaste URI of file:$uri")](https://github.com/ankidroid/Anki-Android/blob/dddac58078ee18236b9b703355bfd159f3311075/AnkiDroid/src/main/java/com/ichi2/anki/MediaRegistration.kt#L154)

This seems to be the origin of error.

[src/main/java/com/ichi2/utils/ContentResolverUtil.kt](https://github.com/ankidroid/Anki-Android/blob/dddac58078ee18236b9b703355bfd159f3311075/AnkiDroid/src/main/java/com/ichi2/utils/ContentResolverUtil.kt)
[`
extension = mime.getExtensionFromMimeType(contentResolver.getType(uri))
`
](https://github.com/ankidroid/Anki-Android/blob/dddac58078ee18236b9b703355bfd159f3311075/AnkiDroid/src/main/java/com/ichi2/utils/ContentResolverUtil.kt#L61)

The mimeType returned is vnd.android.cursor.dir/image for which there seems to be no mapped extension.

This also happens in master (previous implementation) branch for me.
new implementation calls onImagePaste with same uri as in master.

**Comment by david-allison-1** [comment]
Could you relay this to #8880. We haven't had a reproduction case before.

It would be best to resolve the above issue before getting this PR in.

Is there any way to get the MIME type or image name from the image if you look in a debugger, or query the cursor to see what access you have.

**Comment by murugu-21** [comment]
It appears there's an issue opened 13 days ago #10240 describing the same problem.

**Comment by murugu-21** [comment]
I have opened a new pr to fix image paste #10320, someone please review.

**Comment by murugu-21** [comment]
> @murugu-21 you have my sincere apologies this has sat idle - I just looked through #10320 and left comments there, that one should go through really soon I think. Once it merges what will be left to do here? Just rebase after the merge and then it should work? Or is there more?

I hope so.

**Comment by mikehardy** [comment]
Wow! I just looked through all the history on this one. It seems so simple but really did end up being detail-oriented and time-consuming. A lot of people use this though and it will be a big help. Sincere thanks for persisting on this and getting these PRs in

**Comment by murugu-21** [comment]
can confirm image paste works as before with new implementation now in my local testing (one ui 3, android 11)

File: AnkiDroid/src/main/java/com/ichi2/anki/FieldEditText.java [hunk 6]
```diff
@@ -137,51 +142,48 @@ public void setImagePasteListener(ImagePasteListener imageListener) {
...
+                Uri uri = clip.getItemAt(i).getUri();
+                try {
+                    mImageListener.onImagePaste(this, uri);
+                }
+                catch (Exception e) {
```
Comment (mikehardy) [inline AnkiDroid/src/main/java/com/ichi2/anki/FieldEditText.java:176]: Our java style guide is not enforced by a lint check, this is just a spacing suggestion, it has nothing to do with functionality but matches java style here

```suggestion
            for (int i = 0; i < clip.getItemCount(); i++) {
                Uri uri = clip.getItemAt(i).getUri();
                try {
                    mImageListener.onImagePaste(this, uri);
                } catch (Exception e) {
```

**Review by mikehardy — APPROVED** [review]
LGTM, I made a spacing suggestion but it's non-functional, my approve should be considered good even after adding that change
Definitely needs a second pair of :eyes: though :-)

**Comment by mikehardy** [comment]
Oh bummer - this took a conflict before it got a second approval. Sometimes GitHub provides me a way to resolve the conflicts using the web UI but not this time. @murugu-21 are you able to use git rebase locally to pull the changes in from tip of main branch and resolve the conflicts then push your changes back out? I'd love to get this in, and I'm sorry that entropy took hold of it

**Comment by mikehardy** [comment]
This is quite subtle now! Thanks for patching up the conflict, it looks like something about the Kotlin migration here affected part of this class though ! #10536 - not sure exactly what it may not be that PR but that's what came up when I did a scan. These interactions are pretty rare, this PR seems to be throwing in *all* the subtle difficulties with getting a PR done in a large project. What a pain!


```

com.ichi2.anki.DeckPickerTest > checkIfClickOnCountsLayoutOpensStudyOptionsOnMobile[TestingAVD(AVD) - 11] SKIPPED 

com.ichi2.anki.FieldEditLineTest > testSaveRestore[TestingAVD(AVD) - 11] FAILED 
	java.lang.NoSuchMethodError: No virtual method setOrd(I)V in class Lcom/ichi2/anki/FieldEditText$SavedState; or its super classes (declaration of 'com.ichi2.anki.FieldEditText$SavedState' appears in /data/app/~~1SVyQxqx94X4GzRne73LZg==/com.ichi2.anki-jMh2BA75E7HgyP8aOwSRZw==/base.apk!classes13.dex)
	at com.ichi2.anki.FieldEditText.onSaveInstanceState(FieldEditText.kt:185)

com.ichi2.anki.NoteEditorTabOrderTest > testTabOrder[TestingAVD(AVD) - 11] SKIPPED 
```

**Comment by mikehardy** [comment]
Unfortunately our Kotlin migration is actually a semi-automated thing for a really specific reason - it allows us to maintain `git blame` history across the java->kt rename while still having each commit be compilable+testable when you use `git bisect` - if we're going to actually do a Kotlin migraiton on the file it needs to be separate (first maybe?) in 2 commits like this https://github.com/ankidroid/Anki-Android/blob/main/docs/kotlin-migration.md

Might be easier to just handle the change in Java temporarily then migrate later?

**Comment by murugu-21** [comment]
The main branch doesn't have FieldEditText.java since #10536 migrated FieldEditText.java to FieldEditText.kt. When resolving the merge conflict, I added the FieldEditText.java file instead of deleting it, this caused the test failure. Since the file has already been migrated, wouldn't old commits be tracked there? restoring FieldEditText.java would break the migration in #10536. Maybe I close this pr and open a new one so that FieldEditText.java isn't tracked in any of the pr's commits?

**Comment by mikehardy** [comment]
Well, the most correct thing to do is to rebase, remove your edits to the java file during the rebase, and apply the changes to the new Kotlin. This specific kind of rebase is pretty messy though I will be the first to admit, with the file completely going away. If you wanted to close the PR and reapply the changes cleanly in a new one I would understand. I continue to be amazed at the difficulty level of this one, and I really appreciate that you are sticking with it.

**Comment by murugu-21** [comment]
closing this one and creating new pr #10617

**Pull request closed without merging**